### DISCUSSION TITLE
📋📌 v5.0.0 特性征集与讨论 | v5.0.0 discussion

### DISCUSSION THREAD

**Discussion by BernardXiong** [RT-Thread/rt-thread#6224] [Ideas]
v5.0.0是一个大版本发布，一些前期的准备也在推进，考虑到的部分包括：

* 内核相关
  - [x] 线程控制块 隐性继承rt_object的问题  
  - [ ] 对ipc.c进行拆分，这样也可以更清晰；
  - [x] 完善整体的多核代码，特别是内核代码，核心组件（文件系统，网络协议栈，命令行）的多核支持；
    - [x] 对调度器进行拆分，分成公共部分，单核部分，SMP多核部分；
  - [ ]  加入tls（Thread Local Storage）特性，可以在tls基础上加入更多的辅助功能
  - [ ] 消息邮箱消息队列返回消息长度的问题 https://github.com/RT-Thread/rt-thread/issues/5730
- [ ] 更好的Cortex-A支持；
  * 在Cortex-A上启用device tree，fdt代码大部分以软件包方式存在；
- [x] 更好的64位支持；
  * TCP/IP协议栈部分整体验证过，保证64位情况下都没问题；
  * 文件系统相关的都进行验证通过；
- [x] 合并rt-smart分支，尽可能的影响只涉及到`components/lwp`组件的情况即可支持MMU下全进程的特性；
  * 支持用户态下以rust方式编程；
  * 支持用户态的rust egui移植（应用）；
- [x] 集成Flash管理组件，如FAL或类似的；
- [ ] 整理，梳理驱动框架、及bsp驱动命名规则。
  - [ ] rtdef.h中和设备相关的，都尽可能分离到rtdevice.h中；
  - [ ] 实现LLRB和栈数据结构，将ipc文件夹拎出来

其他请多多补充。

**Comment by Guozhanxin** [comment]
https://github.com/RT-Thread/rt-thread/issues/5619

**Comment by Guozhanxin** [comment]
驱动框架、及bsp驱动命名规则整理。

**Comment by mysterywolf** [comment]
LLVM https://github.com/RT-Thread/rt-thread/issues/5551

----

https://github.com/RT-Thread/rt-thread/pull/7521 <- 已经在此PR中得到支持

**Comment by mysterywolf** [comment]
线程控制块 隐性继承rt_object的问题 https://github.com/RT-Thread/rt-thread/pull/5500

-------

已经在此处 完成 https://github.com/RT-Thread/rt-thread/pull/7131

**Comment by armink** [comment]
目前缺少 Flash 管理类的组件，挺多嵌入式产品上都会用到相关功能。存储也算是系统基础能力，是不是可以把 FAL 从软件包移动至组件里。

------------

已完成

**Comment by mysterywolf** [comment]
GCC下完整的自主的libc，不依赖newlib

**Comment by BernardXiong** [comment]
- [x] 对调度器进行拆分，分成公共部分，单核部分，SMP多核部分；
- [ ] 对ipc.c进行拆分，这样也可以更清晰；
- [ ] rtdef.h中和设备相关的，都尽可能分离到rtdevice.h中；
- [ ] 🙋‍♂️ 讨论：纳入nimble stack到发布版本中；

**Reply by mysterywolf** [reply]
不建议nimble stack放在主仓中，到时候又搞的和lwip一样，三个版本一坨

**Reply by heyuanjie87** [reply]
device.c移动到驱动目录

**Comment by mysterywolf** [comment]
ipc 的FIFO模式也有问题： https://github.com/RT-Thread/rt-thread/issues/4940

**Comment by BernardXiong** [comment]
- [ ] 加入tls（Thread Local Storage）特性，可以在tls基础上加入更多的辅助功能

**Reply by mysterywolf** [reply]
赞同，TCB的user_data成员变量应该由TLS取代，否则很容易出现多个上层应用抢一个user_data的问题。

**Comment by mysterywolf** [comment]
更新（2022年8月28日13:39:00）在此处实现 https://github.com/RT-Thread/rt-thread/pull/6340


-----------------

版本控制规范：https://semver.org/

目前的版本子号不是很规范，其中RT_VERSION甚至和VS那边造成了冲突
```c
/* RT-Thread version information */
#define RT_VERSION                      4L              /**< major version number */
#define RT_SUBVERSION                   1L              /**< minor version number */
#define RT_REVISION                     1L              /**< revise version number */
```

参考ESP32：
```c
/** Major version number (X.x.x) */
#define ESP_ARDUINO_VERSION_MAJOR   2
/** Minor version number (x.X.x) */
#define ESP_ARDUINO_VERSION_MINOR   0
/** Patch version number (x.x.X) */
#define ESP_ARDUINO_VERSION_PATCH   3
```

建议改成：
```c
/* Major version number (X.x.x) */
#define RT_VERSION_MAJOR   4
/* Minor version number (x.X.x) */
#define RT_VERSION_MINOR   2
/* Patch version number (x.x.X) */
#define RT_VERSION_PATCH   0
```

**Comment by mysterywolf** [comment]
消息邮箱消息队列返回消息长度的问题：https://github.com/RT-Thread/rt-thread/issues/5730
消息邮箱不返回消息长度可能还能理解一下，因为要保持小巧、简单。但是消息队列应该返回实际消息的长度。

**Comment by jiuxiaxixi** [comment]
USB 方式连接4G 网络？

**Comment by jiuxiaxixi** [comment]
> 目前缺少 Flash 管理类的组件，挺多嵌入式产品上都会用到相关功能。存储也算是系统基础能力，是不是可以把 FAL 从软件包移动至组件里。

好像latest 已经这么做了，不过好像是默认支持STM32F2 这个搞得有点尴尬

**Comment by jiuxiaxixi** [comment]
支持用户态下以rust方式编程； 这个棒！

**Comment by mysterywolf** [comment]
> 好像latest 已经这么做了，不过好像是默认支持STM32F2 这个搞得有点尴尬

那个F2只是个例子，你需要自己适配到自己的具体型号，部分bsp也适配了fal，你可以参考

**Comment by mysterywolf** [comment]
考虑在v5.0.0中多增加几个RT_Exxx的错误码，相比如uCOS-III的错误码我们的错误码确实比较模糊。大部分错误都被囫囵吞枣的归结到了-RT_ERROR这种没有任何实际指导意义的错误码上。

--------

看了一下freertos的，更少。。。

**Comment by jiuxiaxixi** [comment]
> 那个F2只是个例子，你需要自己适配到自己的具体型号，

我的意思是 这个 这个放在 drv_fal.c 里面，放在 rtthread/componets 里面可能有点不合适，应该是和芯片没有关系才是对的

**Comment by mysterywolf** [comment]
考虑将rtdebug.h 和 rtdbg.h整合到成一个文件

**Comment by BernardXiong** [comment]
- [ ] SConscript下加入对xml构建方式支持；

**Comment by esonhon** [comment]
感觉目前components和在线组件等 KConfig菜单结构和组织有些混乱，不太好找。中移的OnesOs应该是基于我们RTT的，但它的菜单结构要条理的多（组件相对较少也是部分原因），可以参考。

**Comment by flyucc** [comment]
> 目前缺少 Flash 管理类的组件，挺多嵌入式产品上都会用到相关功能。存储也算是系统基础能力，是不是可以把 FAL 从软件包移动至组件里。

使用fal时，总是要去编辑fal_cfg.h，这个能改成通过api注册的方式就好了

**Comment by BernardXiong** [comment]
> > 目前缺少 Flash 管理类的组件，挺多嵌入式产品上都会用到相关功能。存储也算是系统基础能力，是不是可以把 FAL 从软件包移动至组件里。
> 
> 使用fal时，总是要去编辑fal_cfg.h，这个能改成通过api注册的方式就好了

这部分如果可以有个UI工具，以图像化方式来创建/编辑是否更好？

**Comment by mysterywolf** [comment]
> 这部分如果可以有个UI工具，以图像化方式来创建/编辑是否更好？

这种方案我之前试验过，可以让sconscript脚本调用一个python脚本，python脚本里使用tkinter绘制图形界面。在执行scons --target=mdk5等操作的时候，这个GUI会自动蹦出来，让用户通过按钮设置，设置的结果返回给sconscript，sconscript再根据设置情况来决定添加什么.c文件或者开启什么宏定义等等。但是路子比较野，所以也就是试验了一下而已。https://github.com/mysterywolf/scons-python

**Comment by Guozhanxin** [comment]
完善 C++ 的支持
- 有一套 cpp 标准库的测试集合
- 文档梳理清楚，明确支持哪些特性，有个清单

**Comment by oyoungs** [comment]
添加cmake构建，这样可以使用其他的IDE(比如CLion, VSCode, QtCreator等)更好的代码编辑， 也可以更方便的对项目构建配置编译宏

**Comment by mysterywolf** [comment]
> 添加cmake构建，这样可以使用其他的IDE(比如CLion, VSCode, QtCreator等)更好的代码编辑， 也可以更方便的对项目构建配置编译宏

CMAKE 构建支持的 scons --target=cmake

**Comment by cazure** [comment]
块设备识别mbr分区和gpt分区功能，并且能有一套管理工具支持，msh命令修改创建分区就好了。mbr功能已经有了，gpt我看到gitee的rt-smart分支有人在提交pr了。

还有fal支持动态修改分区表就好了，最好能在fal中直接支持mbr，gpt。我在想这三个东西都是为了分区而存在的，fal难道不能做个超集把mbr gpt融合进来嘛，哈哈哈。手动狗头

**Comment by BernardXiong** [comment]
> 

后续rt-smart那边会先合并，然后5.0版本会把smart分支都合并进来。

**Comment by mysterywolf** [comment]
![image](https://user-images.githubusercontent.com/34888354/166119757-31bdfd73-774f-4d8c-9f15-0a29a9271a0f.png)
设备框架没有提供init open close的接口，导致用户如果想要做一些构造和析构操作，例如init的时候初始化IO，close的时候将IO恢复到默认状态，关闭对应的时钟低功耗，都不能实现。
https://github.com/RT-Thread/rt-thread/issues/611

**Comment by charlown** [comment]
既然重新加入flash管理组件, FAL组件，要不要规范个OTA框架或者接口，那个stm32通用bootloader挺鸡肋的，能把它开放出来给开发者适配其他芯片么

**Comment by mysterywolf** [comment]
object的名字，可否改为const char*类型，目前每一个object都需要提供默认8字节的名字存储空间，这个名字后续实际上也不会更改，放在ram里是死重，实属浪费的。

**Reply by Yohozzy** [reply]
这里其实还有个可以改善一下的地方 因为32位对齐 这8个字节后面带着两个8字节的成员 
中间空了2个字节一直没用

**Comment by mysterywolf** [comment]
互斥量考虑实现一下优先级天花板协议：
优先级继承协议 -》 自动挡汽车
优先级天花板协议-》手动挡汽车
优先级天花板协议在处理复杂条件下的优先级翻转时可以人工手动控制，行为更受控。

**Comment by armink** [comment]
VS 模拟器上开启网络相关功能后存在编译错误：
- 打开 SAL 有错误
- 打开 LWIP 有错误

**Comment by mysterywolf** [comment]
梳理需要增加的API：
- rt_sem/mq/mb..._find()
- rt_mutex_get_holder()
- rt_sem_get_value() (https://github.com/RT-Thread/rt-thread/pull/5948)
- rt_mb/mq_flush()

**Comment by mysterywolf** [comment]
thread结构体下的user_data不能被我方自己所用 dlmodule和pthread都在偷偷地使用这个字段
设备框架中的user_data字段也有这个问题
如果我方内部有可能使用这个字段，就不能用user_data这个名字

**Comment by mysterywolf** [comment]
彻底清除掉finsh的存在，否则会把用户搞蒙 一会shell 一会msh 一会finsh

**Comment by mysterywolf** [comment]
https://github.com/RT-Thread/rt-thread/issues/6007
https://club.rt-thread.org/ask/question/231413344ffb4032.html 
用户对设备框架层还是不能完全掌控
没有丰富的文档和教学视频 尤其是移植和框架层内部的讲解
还有一个问题就是设备框架层的命名和规范非常杂，需要统一。
[框架命名问题.pptx](https://github.com/RT-Thread/rt-thread/files/8945997/default.pptx)
https://github.com/RT-Thread/rt-thread/pull/6055
https://github.com/RT-Thread/rt-thread/issues/5342#issuecomment-1043247181

**Comment by wxuefei** [comment]
有没有 x64 支持？？

**Comment by mysterywolf** [comment]
> 有没有 x64 支持？？

https://github.com/RT-Thread/rt-thread/tree/master/bsp/x86

**Comment by wxuefei** [comment]
> > 有没有 x64 支持？？
> 
> https://github.com/RT-Thread/rt-thread/tree/master/bsp/x86

这是 x86 32位的吧，x86_64 也支持？

**Comment by mysterywolf** [comment]
https://github.com/RT-Thread/rt-thread/pull/6011#issuecomment-1146573689

**Comment by BernardXiong** [comment]
> > > 有没有 x64 支持？？
> > 
> > 
> > https://github.com/RT-Thread/rt-thread/tree/master/bsp/x86
> 
> 这是 x86 32位的吧，x86_64 也支持？

是的，这个是32位的，欢迎来推进x86 64。

**Comment by foldl** [comment]
NimBLE stack sucks. 千万别。

A few reasons:

1. [Configurations](https://github.com/apache/mynewt-nimble/blob/master/porting/nimble/include/syscfg/syscfg.h) sucks;
2. Organization of headers/sources sucks;
3. [struct os_mbuf](https://github.com/apache/mynewt-nimble/blob/master/porting/nimble/include/os/os_mbuf.h) is the worst design I have ever seen.

**Comment by Zmmfly** [comment]
RT Smart上 asio 有支持计划吗？

**Comment by mysterywolf** [comment]
https://github.com/RT-Thread/rt-thread/pull/6103#issuecomment-1165806040

**Comment by cazure** [comment]
有支持fdcan的计划吗？我现在发现设备驱动框架里面好像还没有对FDCAN数据包的处理，都是支持CAN2.0AB标准的数据包。之前我看到论坛有人发帖说在artpi上实现了fdcan实际上好像也是支持到CAN2.0AB的数据包。
https://club.rt-thread.org/ask/question/01ee2cb8a03be0ab.html
https://club.rt-thread.org/ask/question/5c1416f99e0cea0f.html

**Comment by grissiom** [comment]
话说 can 设备驱动，发送和接收的结构体不一样，能忍？……

**Comment by flyucc** [comment]
rt-thread上有计划ADC/DAC驱动模型的DMA支持嘛？

**Comment by mysterywolf** [comment]
https://club.rt-thread.org/ask/question/a6da7113f233e704.html

**Comment by charlown** [comment]
这里我写一些关于BSP驱动的想法吧。

最近梳理了一下gd32片上flash的情况
1 flash容量 64k/128k/256k, page 1kb/2kb/4kb
2 flash容量分成bank0, bank1, page 1kb/2kb/4kb
3 flash容量分成bank0, bank1, 使用sector
4 flash容量512k, page 8kb
5 flash容量 <= 128kb,  page 1kb

个人把drv flash分成以下系列：
drv_flash_l23x.c
drv_flash_f1x0.c
drv_flash_f10x.c
drv_flash_f20x.c
drv_flash_f3x0.c
drv_flash_f30x.c
drv_flash_f4xx.c
drv_flash_f403.c
drv_flash_e10x.c
drv_flash_e23x.c
drv_flash_e50x.c
drv_flash_c10x.c

对于现在的bsp驱动架构：
drv_flash.c形式
1 同一个文件可能存在几千行代码混杂着大量的宏定义，个人认为过于复杂了
2 不断适配新的芯片有可能频繁改动同一文件，以及为了适配新芯片而加入新的宏定义，频繁改动意味着有可能引入bug风险

drv_flash_xxx.c形式
1 减少改动，减少出现同一个文件修改冲突，以及忘记git pull最新源码造成丢失最新pr的情况
2 利用开源的力量完善驱动，有利于任务分发
比如在开源中国分发完善gd32的drv_flash驱动任务，每个挑战者就可以领取不同系列的drv_flash驱动，加快完善驱动的进程

**Comment by mysterywolf** [comment]
![image](https://user-images.githubusercontent.com/34888354/182523488-c863ad0c-a9ea-49c7-a8be-814e103326b9.png)

**Comment by mysterywolf** [comment]
用deadbeef取代# 从栈顶开始扫描来计算max used，有一些debug工具会使用线程栈底部的空间，导致max used计算错误 一直为99%

**Comment by mysterywolf** [comment]
栈对齐由4字节改成8字节

**Comment by tfx2001** [comment]
在 serial_v2 中，`struct rt_uart_ops` 新增了 `transmit` 函数，当一些按照旧版本串口驱动编写的设备驱动被调用时，会直接调用野指针。建议修改类型名称，直接在编译阶段报错；或者增加断言。

https://github.com/RT-Thread/rt-thread/blob/1f134b8dc46d16bd26a58ac08f430a524d23a45a/components/drivers/serial/serial_v2.c#L470-L480

**Reply by tfx2001** [reply]
暂时修复：#6263

**Comment by RiceChen0** [comment]
ac6的支持，目前发现脚本确实有AC6的影子，但是AC6是与GUN兼容，这与AC5不一致。所以libcpu的.S选择存在漏洞。并且存在打开RT_USING_HEAP之后的,也同样存在不兼容问题。

**Reply by mysterywolf** [reply]
demo https://github.com/RT-Thread/rt-thread/pull/6268

**Reply by RiceChen0** [reply]
这个一个提交是不完整的，libcpu的脚本也需要相对应的修正

**Comment by mysterywolf** [comment]
read write接口根据unix相关资料，应该是ssize_t 也就是有符号类型的可以返回-1，而不是无符号类型的rt_size_t/size_t，该定义错误间接导致了 https://github.com/RT-Thread/rt-thread/pull/6185 埋下了隐患

**Reply by mysterywolf** [reply]
已完成

**Comment by mysterywolf** [comment]
https://github.com/RT-Thread/rt-thread/pull/6235#issuecomment-1216065329 RTC框架日期和时间分成了两个函数，容易在比如2023年12月31日23时59分59秒这种节点出问题

**Comment by mysterywolf** [comment]
实现LLRB和栈数据结构，将ipc文件夹拎出来

**Comment by mysterywolf** [comment]
![image](https://user-images.githubusercontent.com/34888354/187090927-34ab08ba-53df-43ab-bec1-47e018caa36f.png)
是否考虑干掉一个？

**Comment by mysterywolf** [comment]
对滥用inline以及宏函数的行为展开筛查

**Comment by winfenggao** [comment]
usb协议栈进行梳理，大部分时候要完全做到拿来就用；可以的话要开始支持OTG了.

**Comment by heyuanjie87** [comment]
单核情况下定义struct rt_spinlock

**Comment by winfenggao** [comment]
动态添加的device要有机制避免应用层调用异常；比如U盘热拔插的过程中读写文件等；

**Comment by heyuanjie87** [comment]
清理libcpu代码调用list_thread();

**Comment by armink** [comment]
- rtc 框架重构，默认使用软件模拟 RTC ，硬件 RTC 定期自动对时。降低频繁的硬件 RTC 访问次数。
 
如下图所示，通过 J-Trace 可以发现，大型项目中 timegm 、  HAL_get_time 、HAL_get_date 占用系统负荷较高

![image](https://user-images.githubusercontent.com/1734686/194876734-94745e2f-7ed5-4c68-ad31-894ca506e28f.png)

**Reply by mysterywolf** [reply]
正有此意！😁

**Comment by mysterywolf** [comment]
https://club.rt-thread.org/ask/article/0242e9ce774e161c.html

**Comment by Yohozzy** [comment]
想要一个统一性的驱动事件回调机制 可以把rx_indicate tx_complete也统合进来 
有些场景 例如我的蓝牙带着一个connect信号 
当我同步读取的时候 蓝牙断开了 卡死在了rt_device_read
当我异步读取的时候 卡在了等回调 
现在的方法是 rx_indicate 回调传入buffer长度为0 再去检查是否已经断开 写起来很奇怪

**Comment by BernardXiong** [comment]
@Guozhanxin @mysterywolf 

随着v5.0的推进，一些原来预先设定的工作可以开展起来的，例如：
* ipc.c拆分，目前这部分太大了；另外也可以参考smart分支，引进例如 RT_INTERRUPTIBLE，RT_KILLABLE，RT_UNINTERRUPTIBLE的特性；
* 对scheduler.c进行拆分，分成scheduler.c（公共部分），scheduler_up.c，scheduler_mp.c三个部分。

**Comment by linshire** [comment]
可以实现一下控制台二级命令甚至参数的自动补全

**Reply by mysterywolf** [reply]
建议暂时不考虑，目前会导致ROM底噪进一步恶化。

**Comment by mr-cn** [comment]
1. 文档中心可以再完善一下
目前好像文档中心是没有历史版本的，新版本的改动是以文字补充的形式体现的。个人认为文档也可以进行版本控制，新版本有新特性，也可能会导致旧版本的部分接口变化，不同的源码应该对应不同的文档。
文档可以新增一个一级栏目，提供对不同平台以及不同BSP目前支持情况，以及跟具体架构、具体平台相关的内容。没看源码的bsp文件夹之前，真不知道RTT已经适配这么多平台了。这样原本的文档也可以再整理，尽量避免通用特性的文档中出现特定平台的内容。设想中的结构类似如下：

- 硬件支持
  - 概况
  - 指令集
    - risc-v
    - arm
    - ...
  - BSP
    - Cortex-A
      - i.MX family
      - ...
    - Cortex-M
      - STM32 family
        - STM32 相关支持总览
        - stm32f103-atk-nano
        - ...

2. 开发流程可以有更灵活的策略
bugfix的小版本更新可以独立于原有的版本发布节奏，有关键bug被修复后可以更快速的发布新版本。
拿 #6647 这个pr举例，去年底MDK更新后，AC6的libc变化了，导致rtt的libc与之冲突，内核无法通过编译。虽然第一时间就有PR修复了，但是到目前为止还暂时没有新的release发布，目前最新的release-4.1.1还是无法在MDK5.38以上版本通过编译。
可以参考[语义化版本](https://semver.org/lang/zh-CN/)和gitflow。

3. ADC驱动框架还不够灵活
目前的ADC框架还是有些不够灵活，高速采样时缺少DMA和连续转换等支持，也没法很方便的使用自定义采样时间、硬件过采样、外部触发等特殊功能，开发时或多或少都要对原有驱动架构大刀阔斧地修改。不过ADC其实应该是挺难统一抽象的设备了，本身就有很多完全不同的工作模式和应用场景；外设差异也大，不要说不同平台不同架构了，就连STM32不同系列之间都有不小的差异。

4. 图形API和USB协议栈管理
这两条其实还有些为时尚早，开发代价很大，在单片机上的应用价值也不大。不过考虑到RT Thread Smart的存在，不同的GUI应用程序直接控制硬件是不现实的，由OS统一管理是十分有必要的；在已经抽象好的文件系统管理和网卡管理等条件下，USB Host应用其实也有相当的想象力了。

**Comment by Luanuxe** [comment]
请问rt-smart的用户态Rust支持目前情况咋样啊
